include/lomse_horizontal_layer.h

```diff
@@ -0,0 +1,96 @@
+//---------------------------------------------------------------------------------------
+// This file is part of the Lomse library.
+// Lomse is copyrighted work (c) 2010-2021. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification,
+// are permitted provided that the following conditions are met:
+//
+//    * Redistributions of source code must retain the above copyright notice, this
+//      list of conditions and the following disclaimer.
+//
+//    * Redistributions in binary form must reproduce the above copyright notice, this
+//      list of conditions and the following disclaimer in the documentation and/or
+//      other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+// OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+// SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// For any comment, suggestion or feature request, please contact the manager of
+// the project at cecilios@users.sourceforge.net
+//---------------------------------------------------------------------------------------
+
+#ifndef __LOMSE_HORIZONTAL_LAYER_H__
+#define __LOMSE_HORIZONTAL_LAYER_H__
+
+#include "lomse_basic.h"
+
+#include <map>
+#include <vector>
+
+namespace lomse
+{
+
+class GmoShape;
+class VerticalProfile;
+
+class HorizontalLayer
+{
+protected:
+    struct OccupiedSpaceInterval
+    {
+        GmoShape* pShape;
+        LUnits xLeft;
+        LUnits xRight;
+    };
+
+    std::map<LUnits, OccupiedSpaceInterval> m_intervals;
+    LUnits m_xAbsLeft;
```

Comment:
I don't see any problem here and there is no need to change anything. I'm just curious about why you designed it in this way. The OccupiedSpaceInterval struct contains the shape and the occupied interval, but the interval is always from shape->left() to shape->right(). Wouldn't it be simpler to just save the shape and define `m_intervals` as `std::map<LUnits, GmoShape*> m_intervals;` ? By using the struct I don't see much gain in speed and requires more memory. But as the number of objects to align in a system is very small, this is not a real problem. So I'm neither asking for changes nor questioning the design. I'm just asking because I would like to understand better your reasoning. Your implementation of HorizontalLayer using a map is very clever and clean, and probably there is some benefit in using the struct that I'm missing.

Reply:
Yes, you are right, in the current version the structure is not needed. This is actually a leftover from my previous versions of `HorizontalLayer`: initially I thought it would be good to split or adjust boundaries of intervals if some of them overlap with the newly added shape, just to keep track on all shapes added to the layer. In this case interval boundaries could differ from the corresponding shapes' boundaries. However when I checked this with some real scores I realized that this creates a total mess when there are some horizontally overlapping shapes so I went for just removing overlapping shapes from the horizontal layer. Now interval boundaries indeed always equal to the boundaries of the corresponding shapes, and no special structure for an interval is needed. I think I'll remove the structure and replace it with a plain pointer to `GmoShape`, thanks for your comment!

Reply:
Understood. Thank you